### PR TITLE
feat: Add GPU acceleration for ROSA architecture

### PR DESCRIPTION
**Pull request by Juste-Leo2**
## GPU Optimization of the ROSA Architecture

**Note:** An important note on this optimization. To achieve this practical speedup, the algorithm's complexity has been modified. We move from a near-linear (O(T)) method to a **quadratic (O(T²))** one.

**The trade-off is:** This version is extremely fast for reasonably sized models, but it is **not viable** for very long contexts (e.g., T > 16k) where the memory and computation time would become prohibitive.

---

This contribution aims to significantly accelerate the ROSA architecture by eliminating a major bottleneck related to the `rosa` algorithm's implementation.

### 1. The Problem: The CPU Bottleneck

The original implementation of `rosa` is written in pure Python. While functional, it introduces a critical performance issue in a GPU training environment:

*   **Data Transfers:** For each call, tensors must be moved from the GPU to the CPU (`.cpu()`) and the results transferred back to the GPU, which is extremely costly.
*   **Sequential Execution:** The Python code uses loops and dictionaries, operations that cannot be parallelized by the GPU.
*   **Parallelism Breaking:** The GPU has to wait for the CPU to finish its computations multiple times during a single forward pass, nullifying the benefits of hardware acceleration.

This made the original architecture very slow and unscalable for large-scale training or inference.

### 2. The Solution: Parallel Re-implementation on GPU

The solution is to re-implement the logic of the `rosa` algorithm using exclusively PyTorch tensor operations, enabling 100% GPU execution.

The sequential approach of building a Suffix Automaton is replaced by a matrix-based, parallel approach:
1.  **Similarity Matrix:** A `(B, T, T)` matrix is computed to identify where tokens are identical within the batch's sequences.
2.  **Dynamic Programming:** A short loop (very fast on GPU) calculates the length of all possible common suffixes.
3.  **Parallel Search:** A vectorized `argmax` over the matrix instantly finds the best past match for every position in the sequence.
4.  **Prediction:** Predictions are retrieved in parallel using `torch.gather`.

This method is **logically equivalent** to the original but is designed to leverage the GPU's parallel processing power.

#### a) New `rosa_batch_gpu_exact` Function

The CPU-based algorithm was replaced by the following function, which is fully executable on the GPU:
```python
def rosa_batch_gpu_exact(x: torch.Tensor, output_uint8: bool = False) -> torch.Tensor:
    """Massively parallel GPU version of the ROSA algorithm."""
    B, T = x.shape
    device = x.device
    eq = (x.unsqueeze(2) == x.unsqueeze(1))
    L = torch.zeros((B, T, T), device=device, dtype=torch.int16)
    for i in range(1, T):
        prev_diag = F.pad(L[:, i - 1, :-1], (1, 0))
        L[:, i, :] = (prev_diag + 1) * eq[:, i, :]
    mask = torch.tril(torch.ones((T, T), device=device, dtype=torch.bool), diagonal=-1)
    j_indices = torch.arange(T, device=device, dtype=torch.float32)
    tie_breaker = j_indices * 1e-5
    L_scored = L.float() + tie_breaker.unsqueeze(0).unsqueeze(1)
    L_scored.masked_fill_(~mask.unsqueeze(0), -1.0)
    max_vals, best_j = L_scored.max(dim=2)
    pred_idx = (best_j + 1).clamp(max=T - 1)
    predictions = torch.gather(x, 1, pred_idx)
    no_match_mask = (L.masked_fill(~mask.unsqueeze(0), 0).max(dim=2).values == 0)
    predictions[no_match_mask] = -1
    if output_uint8:
        return predictions.clamp_min(0).to(torch.uint8)
    else:
        return predictions.to(x.dtype)
```

#### b) Integration into Model Layers

The calls to the old CPU functions were simply replaced in the `rosa_emb_layer` and `rosa_4bit_layer` layers.

**Example in `rosa_emb_layer`:**
```diff
- idx_pred = rosa_batch_python_orig(idx)
+ idx_pred = rosa_batch_gpu_exact(idx, output_uint8=False)
```

**Example in `rosa_4bit_layer`:**
```diff
- idx_q = rosa_batch_python(tok2d).reshape(B, Cg, T).transpose(1, 2).contiguous()
+ idx_q = rosa_batch_gpu_exact(tok2d, output_uint8=True).reshape(B, Cg, T).transpose(1, 2).contiguous()
```

### 4. Validation and Results

The effectiveness of this optimization was validated by training the model for 1000 steps. Using a learning rate of `3e-3`, we successfully **reproduced the original author's loss curve**, proving that our GPU implementation is functionally correct. (
[train_and_plot_rosa.py](https://github.com/user-attachments/files/22989515/train_and_plot_rosa.py) )

training_loss_plot

Inference with the pre-trained model also confirms that the prediction performance is maintained, with accuracy exceeding 90% on the counting task.

### CODE REVIEW

**Comment by Juste-Leo2** [comment]
I’m going to close this fork and propose a C++ compilation with CUDA to significantly speed up model training 🚀
With this compiled version, I’m getting up to 30x faster performance on the small model 

image
